### ISSUE TITLE
KataContainers/CLH is reserving high PageCache memory with DevMapper

### ISSUE THREAD

**Issue by RamyAllam** [kata-containers/kata-containers#4418]
# Description of problem

I'm running KataContainers/CloudHypervisor with K8s/Containerd on devmapper/direct lvm.  And I'm seeing that devmapper when used by KataContainers is reserving much of the page cache that can be seen as `buff/cache`.

This page cache is not removed by the OS unless
- The Pod containing the data is deleted 
- OR manually dropping the cache using `echo 1 > /proc/sys/vm/drop_caches`

# Expected result
DevMapper on CLH to reserve a lower page cache and don't cache the entire file.  

# Actual result
KataContainers/CloudHypervisor put 15G into the worker node page cache.

# Further information
You can reproduce the issue by running a Pod with CloudHypervisor and devmapper as the storage backend and execute the following inside the Pod
```
dd if=/dev/zero of=/tmp/test$$ bs=1024 count=15680000
```

You will notice that the `buff/cache` on the worker node is incremented by 15G and is kept there.  I thought this can be a devmapper issue, however, when running the Pod directly on containerd/devmapper without KataContainers I don't see that issue and the reserved cache is only a few mBytes.

**Comment by RamyAllam** [comment]
[kata.log](https://github.com/kata-containers/kata-containers/files/8855408/kata.log)
![pod_withKata_afterDD](https://user-images.githubusercontent.com/15165030/172448744-bed30320-a847-4cf2-aa42-6652043eb161.jpg)
![pod_withKata_beforeDD](https://user-images.githubusercontent.com/15165030/172448756-554625e7-5e5a-4a7b-bbb5-9b04c080618d.jpg)
![pod_withoutKata_afterDD](https://user-images.githubusercontent.com/15165030/172448764-64d4183e-792b-4204-9890-7df7b6f1341a.jpg)
![pod_withoutKata_beforeDD](https://user-images.githubusercontent.com/15165030/172448775-804a6d1c-77ff-4e92-9452-8c72c9e929cf.jpg)

**Comment by rbradford** [comment]
Bypassing the page cache can be enabled by setting `direct` to `true` when creating the virtio-block device. This is in `DiskConfig`: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/config.rs#L952

It's not on by default because some filesystems (like tmpfs) don't support it but as Kata will only use this with device mapper it is probably safe for Kata to turn it on all the time (or at least when running with device mapper.)

CC @sboeuf @fidencio

**Comment by sboeuf** [comment]
> Bypassing the page cache can be enabled by setting `direct` to `true` when creating the virtio-block device. This is in `DiskConfig`: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/config.rs#L952
> 
> It's not on by default because some filesystems (like tmpfs) don't support it but as Kata will only use this with device mapper it is probably safe for Kata to turn it on all the time (or at least when running with device mapper.)
> 
> CC @sboeuf @fidencio

I agree it'd be nice to be able to enable this option through Kata. Not sure about enabling it by default since some users might prefer the performance brought by the page cache over the actual savings in terms of memory.
@fidencio is there such knob available from the Kata configuration? Something similar for QEMU that could be reused by Cloud Hypervisor?

**Comment by RamyAllam** [comment]
Thanks everyone 👍 
Is there a workaround to enable "direct" until a new option is introduced to Kata? OR do I need to compile CloudHypervisor instead with "direct" enabled and replace it with the binary that Kata provides?

**Comment by sboeuf** [comment]
I think you can simply add the following after this [line](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/clh.go#L672):
```go
clhDisk.Direct = true
```

**Comment by cjtim** [comment]
I faced similar issues with QEMU as well but i already applied
```
block_device_driver = "virtio-scsi"
block_device_cache_direct = true
```
Anyway to limit the size of it can grow?

**Comment by amshinde** [comment]
@cjtim According to @sboeuf's suggestion the line needs to be changed in the code : https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/clh.go#L672 with adding
```clhDisk.Direct = true```
Can you check if that works for you?

**Comment by amshinde** [comment]
@RamyAllam  Did the suggestion work for you to use `Direct` for the block device? I am surprised as based on your initial comments you are writing to /tmp/test which should not go to rootfs backed block device, as you are essentially writing to tmpfs there. 
Curious, if you saw similar issue with in non-devicemapper case as well.

**Comment by cjtim** [comment]
> @cjtim According to @sboeuf's suggestion the line needs to be changed in the code : https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/clh.go#L672 with adding `clhDisk.Direct = true` Can you check if that works for you?

I use QEMU VMM. 😅

**Comment by skaegi** [comment]
Even with...
```
clhDisk.Direct = func(b bool) *bool { return &b }(true)
```
... hard-coded I'm seeing the buff/cache grow "however" it eventually seems to cap. Without `Direct` set it grows until there is 0 free -- albeit it's all still available.
